### PR TITLE
Replace usages of InterruptPending to the flag of query cancellation

### PR DESCRIPTION
**Pull request by maksm90**
This PR fixes https://github.com/greenplum-db/gpdb/issues/13048.

Currently some communication routines between master and segments performing polling of messages from segments rely on `InterruptPending` flag in detection of cancel or termination events [[1](https://github.com/greenplum-db/gpdb/blob/a0399104229dda9099bcb5673e8b9d1a5ae3ac96/src/backend/cdb/dispatcher/cdbdisp_async.c#L484-L490),[2](https://github.com/greenplum-db/gpdb/blob/a0399104229dda9099bcb5673e8b9d1a5ae3ac96/src/backend/cdb/dispatcher/cdbdisp_query.c#L1162-L1163),[3](https://github.com/greenplum-db/gpdb/blob/a0399104229dda9099bcb5673e8b9d1a5ae3ac96/src/backend/cdb/cdbcopy.c#L524-L527),[4](https://github.com/greenplum-db/gpdb/blob/a0399104229dda9099bcb5673e8b9d1a5ae3ac96/src/backend/cdb/motion/ic_tcp.c#L2374-L2380),[5](https://github.com/greenplum-db/gpdb/blob/a0399104229dda9099bcb5673e8b9d1a5ae3ac96/src/backend/cdb/motion/ic_tcp.c#L2396-L2397)]. This flag is used to define some interruption in CHECK_FOR_INTERRUPTS() routine. Up to some time this flag was closely assigned with some exceptional case, more often with cancel or termination request. But as GPDB merges with newest versions of PostgreSQL, there are more and more cases where interruption begins to denote some not certainly causing error thing, e.g., periodic check of connectivity with client. For this reason it becomes necessary to redefine the usages of `InterruptPending` flag to no longer erroneously evaluate the set `InterruptPending` flag as cancel request.

Current patch touches four places of `InterruptPending` usage in waiting loops of master-segments communication:

- `checkDispatchResult()` (in blocking mode) blocks QD flow [until](https://github.com/greenplum-db/gpdb/blob/a0399104229dda9099bcb5673e8b9d1a5ae3ac96/src/backend/cdb/dispatcher/cdbdisp_async.c#L492-L547) all QEs report some response or an error. The blocking might last a long time especially when all execution is performed on QEs' side, e.g. for `INSERT INTO ... SELECT ...` QD spends all its time in `checkDispatchResult()` function within `mppExecutorFinishup()` call. Therefore, it’s helpful to integrate CFI call in waiting loop and define cancel event based on it.
- `cdbdisp_dispatchX()` allows to cancel dispatching of plan slices somewhere [in the middle](https://github.com/greenplum-db/gpdb/blob/a0399104229dda9099bcb5673e8b9d1a5ae3ac96/src/backend/cdb/dispatcher/cdbdisp_query.c#L1155-L1164). All dispatching happens in non-blocking mode so it’s enough to react only on cancel/termination requests. All other signals will be handled further without delay in dispatching loop.
- `waitOnOutbound()` is called by `TeardownTCPInterconnect()` function to [receive](https://github.com/greenplum-db/gpdb/blob/a0399104229dda9099bcb5673e8b9d1a5ae3ac96/src/backend/cdb/motion/ic_tcp.c#L2415-L2421) STOP message from segments [before](/https://github.com/greenplum-db/gpdb/blob/a0399104229dda9099bcb5673e8b9d1a5ae3ac96/src/backend/cdb/motion/ic_tcp.c#L2415-L2421) closing connections. We might hang on waiting loop for a significant period of time when `select()` completes by timeout on each iteration of waiting loop, i.e. when connectivity is kept but segments delay the sending of STOP message or when master as tcp client cannot detect disconnection of network (`tcp_keepalive` has to repair it). All of these are rare enough so regular check on cancel/termination is enough.
- `cdbCopyInternal()` is called in COPY TO/FROM operations issued on master after data transferring. Exit from [waiting loop](https://github.com/greenplum-db/gpdb/blob/a0399104229dda9099bcb5673e8b9d1a5ae3ac96/src/backend/cdb/cdbcopy.c#L522-L533) occurs after receiving of responses from segments or when connection has failed (`PQstatus() != CONNECTION_OK`). Hanging is possible in all cases as in `waitOnOutbound()` function. Simple cancel/termination check is enough.

### CODE REVIEW

**Comment by ashwinstar** [comment]
Thanks @maksm90 for this PR, team will review it soon

**Comment by interma** [comment]
@maksm90 Thanks for your PR! And took a first look, this PR is good.

Besides you mentioned, there are 3 other places which `InterruptPending` be used.
1. https://github.com/greenplum-db/gpdb/blob/9b451aac10c8ab257e4c2cb4a63efa23c8c4980f/src/backend/commands/vacuum.c#L2594
2. https://github.com/greenplum-db/gpdb/blob/9b451aac10c8ab257e4c2cb4a63efa23c8c4980f/src/backend/utils/mmgr/vmem_tracker.c#L564
3. https://github.com/greenplum-db/gpdb/blob/9b451aac10c8ab257e4c2cb4a63efa23c8c4980f/src/backend/regex/regcomp.c#L2021

I think 1&2 is no need to change, and 3 can be refactored to `CancelRequested()` (in your PR). Just my option, what's your idea?

And continue to review today: main focus on cdb dispatch logic.

File: src/backend/cdb/dispatcher/cdbdisp_async.c [hunk 1]
```diff
@@ -482,11 +474,27 @@ checkDispatchResult(CdbDispatcherState *ds, int timeout_sec)
...
-		if ((InterruptPending || meleeResults->errcode) && meleeResults->cancelOnError)
+		PG_TRY();
+		{
+			CHECK_FOR_INTERRUPTS();
+		}
+		PG_CATCH();
+		{
+			cancelRequested = true;
+		}
+		PG_END_TRY();
```
Comment (interma) [inline src/backend/cdb/dispatcher/cdbdisp_async.c:489]: Why no `CancelRequested()` needed here? I suppose the logic may like
```
CHECK_FOR_INTERRUPTS();
if (CancelRequested())
    cancelRequested = true;
```

Thanks.

Comment (interma) [reply src/backend/cdb/dispatcher/cdbdisp_async.c:489]: got it!

**Comment by interma** [comment]
@maksm90 thanks for your explanation. 
+1 for your PR, thanks for your contribution again!

And if this PR is merged, there is no other blocker for https://github.com/greenplum-db/gpdb/pull/12223, right?

**Review by interma — APPROVED** [review]
LGTM

**Comment by maksm90** [comment]
@interma, thx for review.

> And if this PR is merged, there is no other blocker for #12223, right?

Yeah, it's so.

**Comment by z-wenlin** [comment]
This commit should be shipped with https://github.com/greenplum-db/gpdb/pull/12223, right? I saw that https://github.com/greenplum-db/gpdb/pull/12223 is still in draft status, will these two commits be merged together?

Besides, will these commits also backport to 6X_STABLE?

**Comment by maksm90** [comment]
> This commit should be shipped with #12223, right?

Not necessarily.

> I saw that #12223 is still in draft status, will these two commits be merged together?

In source code current change doesn't intersect with that one presented in #12223 but correctness of work of the second one directly depends on the first one. So PR #12223 will be active after merging current one.

> Besides, will these commits also backport to 6X_STABLE?

Yeah, sure

**Comment by maksm90** [comment]
Resolved conflicts

**Comment by interma** [comment]
@maksm90 After all conflicts have been resolved, please help to backport this PR to 6X.
Plan to merge them in the next week.

Thanks!

**Pull request merged**